### ISSUE TITLE
Warning in debug mode when combining 'multiple' field with subpalette

### ISSUE THREAD

**Issue by dennisbohn** [contao/contao#5107]
### Affected version(s)

4.13.6

### Description

While testing a feature request for subpalettes and saw the following check for the option 'multiple' in the DC_Table.php.

https://github.com/contao/contao/blob/12c7f37523f4f4bfedaaccb1b322ade711dab52c/core-bundle/src/Resources/contao/drivers/DC_Table.php#L3321

So I tested 'multiple' in combination with a subpalette.

```php
'home' => [
    'inputType'              => 'checkbox',
    'eval' => [
        'submitOnChange'     => true,
        'multiple'           => true
    ],
    'options' => [
        '1'                  => 'a',
        '2'                  => 'b',
        '3'                  => 'c'
    ],
    'sql'                    => "varchar(255) NOT NULL default ''",
]
```

I got the following warning in debug mode when the script tries to combine a string with an array.

![image](https://user-images.githubusercontent.com/87128053/183776419-9c6cb0be-77c6-4fd8-8440-75c06f5c0301.png)

https://github.com/contao/contao/blob/12c7f37523f4f4bfedaaccb1b322ade711dab52c/core-bundle/src/Resources/contao/drivers/DC_Table.php#L3334

**Comment by aschempp** [comment]
Well imho you can't use a multiple-Checkbox to trigger a subpalette, so this is an "intended" error due to a misconfiguration. Or what would you fix?

**Comment by dennisbohn** [comment]
Ok, I was wondering, why the if statement exists when multiple wasnt intended to be used with subpalettes. I suggested a change for subpalettes with selects and wasnt shure if that could cause a problem in other cases.

https://github.com/contao/contao/issues/5106

**Comment by dennisbohn** [comment]
Maybe it would be clearer when the if statement wraps around the if/else.

```php
if ($trigger && !($GLOBALS['TL_DCA'][$this->strTable]['fields'][$name]['eval']['multiple'] ?? null))
{
	if (($GLOBALS['TL_DCA'][$this->strTable]['fields'][$name]['inputType'] ?? null) == 'checkbox')
	{
		$sValues[] = $name;

		// Look for a subpalette
		if (isset($GLOBALS['TL_DCA'][$this->strTable]['subpalettes'][$name]))
		{
			$subpalettes[$name] = $GLOBALS['TL_DCA'][$this->strTable]['subpalettes'][$name];
		}
	}
	else
	{
		$sValues[] = $trigger;
		$key = $name . '_' . $trigger;

		// Look for a subpalette
		if (isset($GLOBALS['TL_DCA'][$this->strTable]['subpalettes'][$key]))
		{
			$subpalettes[$name] = $GLOBALS['TL_DCA'][$this->strTable]['subpalettes'][$key];
		}
	}
}
```

Or to combinate it with my suggestion to allow 0 with selects. (https://github.com/contao/contao/issues/5106)

```php
if (!($GLOBALS['TL_DCA'][$this->strTable]['fields'][$name]['eval']['multiple'] ?? null))
{
	if (($GLOBALS['TL_DCA'][$this->strTable]['fields'][$name]['inputType'] ?? null) == 'checkbox')
	{
		if ($trigger)
		{
			$sValues[] = $name;

			// Look for a subpalette
			if (isset($GLOBALS['TL_DCA'][$this->strTable]['subpalettes'][$name]))
			{
				$subpalettes[$name] = $GLOBALS['TL_DCA'][$this->strTable]['subpalettes'][$name];
			}
		}
	}
	else
	{
		$sValues[] = $trigger;
		$key = $name . '_' . $trigger;

		// Look for a subpalette
		if (isset($GLOBALS['TL_DCA'][$this->strTable]['subpalettes'][$key]))
		{
			$subpalettes[$name] = $GLOBALS['TL_DCA'][$this->strTable]['subpalettes'][$key];
		}
	}
}
```

**Comment by dennisbohn** [comment]
It seems, a workaround isnt possible at the moment.

I tried to used the callback "fields.home.options" to generate a select with "none" instead of "0", used callback "fields.home.save" to convert "none" to "0" when writing to database and used callback "fields.home.load" to convert "0" back to "none" when reading from database.

Unfortunately the load callback is completely ignored because the value comes directly from the database.

https://github.com/contao/contao/blob/12c7f37523f4f4bfedaaccb1b322ade711dab52c/core-bundle/src/Resources/contao/drivers/DC_Table.php#L3297-L3299

**Comment by leofeyer** [comment]
@dennisbohn Has this been fixed with #5112?

**Comment by dennisbohn** [comment]
You can still produce the warning when you misconfigure the dca (multiple field + subpalette). I was not sure if the error was intended so I changed as litte as possible in the pull request just to allow the suffix "_0" in subpalettes.

When a multiple field was combined with a subpalette the variable ($trigger) contained an array.

Because of the if statement that checked, if the field is not a multiple field you ran into the following code.

```php
$sValues[] = $trigger;
$key = $name . '_' . $trigger;

// Look for a subpalette
if (isset($GLOBALS['TL_DCA'][$this->strTable]['subpalettes'][$key]))
{
	$subpalettes[$name] = $GLOBALS['TL_DCA'][$this->strTable]['subpalettes'][$key];
}
```

In the definition of $key the script tries to combine the string ($name) with the array ($trigger) and you get the warning. I can create a new slightly modified pull requests to skip this block completely in case of a multiple field.

```php
if (!($GLOBALS['TL_DCA'][$this->strTable]['fields'][$name]['eval']['multiple'] ?? null))
{
	if (($GLOBALS['TL_DCA'][$this->strTable]['fields'][$name]['inputType'] ?? null) == 'checkbox')
	{
		if ($trigger)
		{
			$sValues[] = $name;

			// Look for a subpalette
			if (isset($GLOBALS['TL_DCA'][$this->strTable]['subpalettes'][$name]))
			{
				$subpalettes[$name] = $GLOBALS['TL_DCA'][$this->strTable]['subpalettes'][$name];
			}
		}
	}
	else
	{
		$sValues[] = $trigger;
		$key = $name . '_' . $trigger;

		// Look for a subpalette
		if (isset($GLOBALS['TL_DCA'][$this->strTable]['subpalettes'][$key]))
		{
			$subpalettes[$name] = $GLOBALS['TL_DCA'][$this->strTable]['subpalettes'][$key];
		}
	}
}
```

Modifying the contao code is still a rollercoaster between the wish to help and the fear of destroying the code. 😅
